### PR TITLE
[FLINK-3755] Introduce key groups for key-value state to support dynamic scaling

### PR DESCRIPTION
**Pull request by StefanRRichter**
This pull request is a followup to the preliminary pull request #2376 and addresses all issues subsumed under [FLINK-3755]

In addition to the changes from PR #2376, this PR adds:

# 1) Refactoring of Key Value State

Before, `AbstractStateBackend` was responsible both for checkpointing to streams and
for keyed state. Now, this functionality is split up into `CheckpointStreamFactory` and
`KeyedStateBackend`. The former is responsible for providing streams for writing checkpoints
while the latter is only responsible for keeping keyed state. A `KeyedStateBackend` can
write its content to a checkpoint. For this it uses a `CheckpointStreamFactory`.

# 2) Introduction of key-group aware `KeyedStateBackend`s

## a) HeapKeyedStateBackend
`HeapKeyedStateBackend` subsumes the keyed state part of both `MemStateBackend` and 
`FsStateBackend` and `MemoryStateBackend`. The only difference between the two now
is that one produces a `CheckpointStreamFactory` that produces streams for writing to files
while the other provides streams that write to in-memory byte arrays.

Also, this introduces another layer of lookup in the `HeapKeyedStateBackend` to accomodate
storing state per key group. Upon checkpointing the data is written out in a format that
is very similar to the new RocksDB backend. We should make these 100 % compatible as
a follow up.

## b) RocksDBKeyedStateBackend

The RocksDB state backend is now also key-group aware. This happens through an additional 
1-2 byte key-group prefix that is added to each key. On snapshots, the key value states 
for different key-groups are combined through `RocksDBMergeIterator`. All snapshots from
this backend are now running fully asynchronous using an implimentation of
`AbstractAsyncIOCallable`.

# Refactoring of asynchrounous snapshot facilities

Snapshots are now driven by `AsyncCheckpointRunnable`s in `StreamTask`, which are executed 
through a threadpool. `AsyncCheckpointRunnable` is created with a 
`RunnableFuture<KeyGroupsStateHandle>` that is obtained from `KeyedStateBackend` through

```
public abstract RunnableFuture<KeyGroupsStateHandle> snapshot(
		long checkpointId,
		long timestamp,
		CheckpointStreamFactory streamFactory) throws Exception;
``` 

# Review comments on #2376

From the comments on this PR, which can be found under PR #2376, we introduced the following changes:

- In comparison to PR #2376, we dropped the KeyGroupAssigner interface in favor of static methods. 
The reason for this is that the code relies on a consistent key group assignment in several places.
- By default, the max parallelism is chosen as 
```
Math.min(128 , roundToNextPowerOfTwo(parallelism + parallelism / 2))
```
- No blocking on the termination of async snapshot threads.
- Reduced logging.

# Limitiations

Currently, queryable state is not key-group aware and `QueryableStateITCase` is ignored. This will 
be solved in a followup work.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyGroupRangeAssignment.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	public static KeyGroupRange computeKeyGroupRangeForOperatorIndex(
+			int maxParallelism,
+			int parallelism,
+			int operatorIndex) {
+		Preconditions.checkArgument(parallelism > 0, "Parallelism must not be smaller than zero.");
+		Preconditions.checkArgument(maxParallelism >= parallelism, "Maximum parallelism must not be smaller than parallelism.");
+		Preconditions.checkArgument(maxParallelism <= (1 << 15), "Maximum parallelism must be smaller than 2^15.");
+
+		int start = operatorIndex == 0 ? 0 : ((operatorIndex * maxParallelism - 1) / parallelism) + 1;
+		int end = ((operatorIndex + 1) * maxParallelism - 1) / parallelism;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyGroupRangeAssignment.java]: Can't we simplify this expression to
```
int start = (operatorIndex * maxParallelism) / parallelism;
int end = ((operatorIndex + 1) * maxParallelism) / parallelism - 1;
```

**Comment by StefanRRichter** [comment]
I don't think that this is giving us the correct inverse for ```computeOperatorIndexForKeyGroup(...)```. Our test ```CheckpointCoordinatorTest::testCreateKeyGroupPartitions()```generates counter-examples.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyGroupRangeAssignment.java (hunk 1 above)
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyGroupRangeAssignment.java]: I don't think that this is giving us the correct inverse for computeOperatorIndexForKeyGroup(...). Our test CheckpointCoordinatorTest::testCreateKeyGroupPartitions()generates counter-examples.

Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyGroupRangeAssignment.java]: Ah I see. Now it makes sense :-)

**Pull request closed without merging**